### PR TITLE
Split Passes.h into 6 per target devices (NFC)

### PR DESCRIPTION
**Pull request by grypp**
As IREE continued to support more targets, `Passes.h` had grown. This PR splits it per target device.

This PR adds new headers per target device, each implementation includes related header
- CommonPasses.h
- LLVMCPUPasses.h
- LLVMGPUPasses.h
- SPIRVPasses.h
- VMVXPasses.h
- WGSLPasses.h

### CODE REVIEW

**Comment by grypp** [comment]
@MaheshRavishankar, I've implemented the changes we discussed yesterday. Let me know if this is the kind of thing you had in mind. 

I'm open to making additional improvements if necessary. For instance, I could ensure that all passes in LLVMCPU only include `LLVMCPUPasses.h`.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Nice! could we make the passes in each file alphabetical, and also move LLVMCPUPasses.h into LLVMCPU, LLVMGPUPasses.h into LLVMGPU, etc.?

**Comment by MaheshRavishankar** [comment]
> I'm open to making additional improvements if necessary. For instance, I could ensure that all passes in LLVMCPU only include `LLVMCPUPasses.h`.

This would be nice! LLVMCPU passes should only include LLVMCPUPasses.h and CommonPasses.h. If they arent we know we have a layering violation. I suspect VMVX will include LLVMCPUPasses.h in addition to VMVXPasses.h and CommonPasses.h, but thats OK.

We should probably wait for next week to land this change! Thanks a lot for doing this!

**Review by kuhar — COMMENTED** [review]
I wonder if we should consider dropping (some of) these pass creation functions and let tablegen autogenerate them. This happens automatically when we omit the `let constructor` in pass definition.

Are there some tangible benefits to declaring and defining those manually, especially when there are no arguments?

**Comment by ScottTodd** [comment]
> I wonder if we should consider dropping (some of) these pass creation functions and let tablegen autogenerate them. This happens automatically when we omit the `let constructor` in pass definition.
> 
> Are there some tangible benefits to declaring and defining those manually, especially when there are no arguments?

I think those files were written before passes could be declared in tablegen (or at least before that became the standard way). That would be good cleanup too.

**Comment by MaheshRavishankar** [comment]
> I wonder if we should consider dropping (some of) these pass creation functions and let tablegen autogenerate them. This happens automatically when we omit the `let constructor` in pass definition.
> 
> Are there some tangible benefits to declaring and defining those manually, especially when there are no arguments?

Yes, thats part of what I want to do as a follow up. But doing it mondo big file is hard. You need to use a `#ifdef` to enable definition of these constructors, and is easier to do on smaller file lists...

**Comment by kuhar** [comment]
Incremental cleanups make a lot of sense to me.

**Comment by grypp** [comment]
> Nice! could we make the passes in each file alphabetical, 

I do not know how to do this automatically. Can clang-format do that? 

> and also move LLVMCPUPasses.h into LLVMCPU, LLVMGPUPasses.h into LLVMGPU, etc.?

I moved Passes headers to each folder, now PR touches many files.

File: compiler/src/iree/compiler/Codegen/Passes.h [hunk 1]
```diff
@@ -20,6 +20,14 @@
 #include "mlir/Pass/PassOptions.h"
 #include "mlir/Support/LLVM.h"
 #include "mlir/Transforms/DialectConversion.h"
+//===---------------------------------------------------------------------===//
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Passes.h]: Do we still need this file?

**Review by MaheshRavishankar — APPROVED** [review]
This looks fine to land... but there are a couple of more follow ups to be done here.
1) The `Passes.td` file needs to be split as well. Let me know if you are planning to do that too.. if not, Ill take a stab at it.
2) Nit: Passes should be made alphabetical in each header file. There is no automated way to do it, but just have to do it manually once.

This is a great start! Thanks for doing this.

**Review by antiagainst — APPROVED** [review]

**Review by dcaballe — APPROVED** [review]
Love this! Thanks a lot! I was actually thinking about this the other day.

**Comment by antiagainst** [comment]
Hi @grypp, any chance we can address the issues and land this soon? I've a few patches that are blocked by this. Thanks. :)

**Comment by grypp** [comment]
Ordering passes alphabetically and fixing the bazel build problem took some time but I finally landed it

File: compiler/src/iree/compiler/Codegen/Passes.h (hunk 1 above)
Comment (grypp) [reply compiler/src/iree/compiler/Codegen/Passes.h]: I left this file with a single declaration. If you have place to put this function declaration, I can do it in a follow-up PR and remove this file completely

**Comment by MaheshRavishankar** [comment]
Thank you very much @grypp !

**Pull request merged**